lib/server.js

```diff
@@ -0,0 +1,198 @@
+    stderr: true,
+    stream: true,
+  };
+  var exec;
+  container.exec(options).then((_exec) => {
+    exec = _exec;
+    return exec.start(attachoptions);
+  }).then((execStream) => {
+    //handling output
+    var strout = through(function (data) {
+      dbgWrite('out', data);
+      var buf = muxBuffer(msgcode.stdout, new Buffer(data));
+      this.queue(buf);
+    });
+
+    var strerr = through(function (data) {
+      dbgWrite('err', data);
+      var buf = muxBuffer(msgcode.stderr, new Buffer(data));
+      this.queue(buf);
+    });
+
+    exec.modem.demuxStream(execStream, strout, strerr);
+    var strbuf = through();
+    strbuf.pipe(through((data) => {
+      socket.send(data, {binary: true});
+    }));
+    strout.pipe(strbuf);
+    strerr.pipe(strbuf);
+
+    //handling input
+
+    function paused (message) {
+      var buf = new Buffer(new Uint8Array(message));
+      debug('message recieved during pause');
+      debug(buf);
+      switch (message[0]) {
+        case msgcode.resume:
+          strbuf.resume();
+          socket.removeListener('message', paused);
+          socket.on('message', resumed);
+          debug('resumed');
+          break;
+
+        case msgcode.stdin:
+          if (!execStream.write(buf.slice(1), {binary: true})) {
+            strbuf.write(new Buffer([msgcode.pause]), {binary: true});
+            debug('paused');
+          }
+          // debug('server needs drain: %s',!tf);
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+
+    function resumed (message) {
+      var buf = new Buffer(new Uint8Array(message));
+      debug(buf);
+      dbgWrite('IN', buf);
+      switch (message[0]) {
+        case msgcode.pause:
+          strbuf.pause();
+          socket.removeListener('message', resumed);
+          socket.on('message', paused);
+          debug('paused');
+          break;
+
+        case msgcode.stdin:
+          if (!execStream.write(buf.slice(1), {binary: true})) {
+            strbuf.write(new Buffer([msgcode.pause]), {binary: true});
+            debug('paused');
+          }
+          // debug('server needs drain: ',!tf);
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+    socket.on('message', resumed);
+
+    socket.on('disconnect', () => {
+      //should be how to ctrl+c ctrl+d, might be better way to kill
+      execStream.end('\x03\x04\r\nexit\r\n');
+    });
+
+    //start recieving client output again
+    execStream.on('drain', () => {
+      strbuf.write(new Buffer([msgcode.resume]), {binary: true});
+      debug('resumed');
+    });
+
+    //When the process dies, the stream ends
+    execStream.on('end', () => {
+      exec.inspect().then((data) => {
+        var exitBuf = new Buffer([data.ExitCode]);
+        strbuf.write(muxBuffer(msgcode.stopped, exitBuf));
```

Comment:
Ideally the class also has a reference to the `DockerExecWebsocketServer` instance that created it... so that on disconnect it can register it... We might want to add a maximum concurrent connection limit option to `DockerExecWebsocketServer`.